### ISSUE TITLE
Permission constraints to endpoints are not applied over API when modifying record

### ISSUE THREAD

**Issue by MeganerdDev** [nautobot/nautobot#5780]
### Environment
* Nautobot version (Docker tag too if applicable): v1.6.23
* Python version: 3.11
* Database platform, version: pgsql 15
* Middleware(s): N/A

### Steps to Reproduce
1. Create permission for /ipam/prefixes endpoint with VIEW/ADD/CHANGE [actions](https://docs.nautobot.com/projects/core/en/v1.6.19/administration/permissions/#actions), assign to test user
2. Create permission for /extras/tags endpoint with VIEW actions and a [constraint](https://docs.nautobot.com/projects/core/en/v1.6.17/administration/permissions/#constraints) applied, assign to test user
 - example constraint: `{"slug__icontains": "test-whitelisted-tag"}` - should only allow the user to object `test-whitelisted-tag` and not any other tag. Which this does work in API and UI as expected.
3. Send a POST request to update a /ipam/prefixes object with the tag by UUID

### Expected Behavior
Expecting HTTP-403 error, same as if we removed the /extras/tags permission altogether and tried to do the same record update

### Observed Behavior
The record is updated with a HTTP-200 response

### Other notes

* From the UI on object, when you go to edit and open the drop-down widget for Tags does not populate. This obeys the permission constraints as expected.

**Comment by MeganerdDev** [comment]
[Prefix serializer](https://github.com/nautobot/nautobot/blob/bc22ef3fa83e9841e9a5868774eda59fc0e4fd6e/nautobot/ipam/api/serializers.py#L219) validation not pulling restricted filter query to stop the update?
Which seems referenced [here](https://github.com/nautobot/nautobot/issues/5120#issuecomment-1904759343) as well

**Comment by glennmatthews** [comment]
Can you clarify, is this issue with `/ipam/prefixes/` or with `/api/ipam/prefixes/` or both?

**Comment by MeganerdDev** [comment]
Hi @glennmatthews Apologies, seems to only be affecting `/api/ipam/prefixes/`

I checked `/ipam/prefixes/` using a proxy to modify the request and the server response indicates it is preventing the action per the constraint.

I also checked other api endpoints like `/api/dcim/devices/` which also appear to be affected by this bug

**Comment by MeganerdDev** [comment]
For the constraints I also tested an additional scenario originally I should include.

Instead of applying the permission constraint to the extras/tag object directly (from the report) which would prevent access to the specific tag object. I tried applying a constraint on the ipam/prefixes permission that applies if a specific tag is tag assigned to the prefix object. From my perspective I should not also be able to assign a tag? (eg: when you assign a tag to the object it will then 'disappear' from the user)

For this I used a similar whitelist test scenario:
`{"tags__slug__icontains": "test-whitelist-tag"}`

**Comment by lampwins** [comment]
Let's investigate a few things here. We assume this is universal to M2M relationships, but might also apply to FKs. Additionally, need to test if this is API/UI specific.

**Comment by HanlinMiao** [comment]
Hi @MeganerdDev, I tested this on our 1.6 branch with the exact setup as described in the issue, I am not really seeing any abnormal behaviors.
In the issue description, step 3.
"Send a POST request to update a /ipam/prefixes object with the tag by UUID".
Did it mean "PATCH" request? (As POST method is not allowed)
If so, this is the response I got which seems reasonable
```
nautobot-FxFJUkoS-py3.11 ❯ curl -X PATCH \                     
-H "Authorization: Token c8c98037de64fa76a9fb5ac838b731ef52316ae4" \
-H "Content-Type: application/json" \
-H "Accept: application/json; version=1.6; indent=4" \
http://localhost:8080/api/ipam/prefixes/1eab1235-2c7f-4806-a8ad-1449a2f8367e/ \
--data '{"tags": ["8c626504-5cff-417f-bcda-0c52860651fa"]}'
{
    "tags": [
        [
            "Related object not found using the provided ID: 8c626504-5cff-417f-bcda-0c52860651fa"
        ]
    ]
}
~/Desktop/nautobot ltm-1.6
nautobot-FxFJUkoS-py3.11 ❯ curl -X POST \ 
-H "Authorization: Token c8c98037de64fa76a9fb5ac838b731ef52316ae4" \
-H "Content-Type: application/json" \
-H "Accept: application/json; version=1.6; indent=4" \
http://localhost:8080/api/ipam/prefixes/1eab1235-2c7f-4806-a8ad-1449a2f8367e/ \
--data '{"tags": ["8c626504-5cff-417f-bcda-0c52860651fa"]}'
{
    "detail": "Method \"POST\" not allowed."
}
```
Is there anything in the setup that I am misunderstanding?

**Comment by HanlinMiao** [comment]
Testing script
```
"""
Concept testing for NTC-1675
    Can we assign a tag to a device without the proper permissions over API (and not UI)
    UI probably filters in ORM
    API filtering could be a delta, lets find out
"""

import pynautobot

nautobot_url = "http://127.0.0.1:8080"
api_token = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"

nautobot = pynautobot.api(nautobot_url, token=api_token)

device_name = "test-device"
device = nautobot.dcim.devices.get(name=device_name)

if not device:
    print(f"Device '{device_name}' not found.")
    exit(1)

tag_name = "test-tag"
tag = nautobot.extras.tags.get(name=tag_name)

if not tag:
    print(f"Tag '{tag_name}' not found.")


tags = device.tags or []

# I am manually entering the UUID of the 'test-tag' object
# which should not be an object this user can touch?
tags.append("84a8d70b-9e3b-4d5e-bfe0-6ca065fea878")

device_update = device.update({
    "tags": tags
})

if device_update:
    print(f"Tag '{tag_name}' successfully added to device '{device_name}'.")
else:
    print(f"Failed to add tag '{tag_name}' to device '{device_name}'.")
```

**Comment by glennmatthews** [comment]
Confirmed under `next`:

```
>>> device = nautobot.dcim.devices.get(name=device_name)
>>> device
<pynautobot.models.dcim.Devices ('Firewall-3') at 0x1036eaee0>
>>> device.tags
[]
```

```
14:46:57.864 INFO    django.server :
  "GET /api/dcim/devices/?name=Firewall-3 HTTP/1.1" 200 2033
```

```
>>> nautobot.extras.tags.all()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/endpoint.py", line 111, in all
    return response_loader(req.get(), self.return_obj, self)
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/query.py", line 361, in get
    return req_all()
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/query.py", line 319, in req_all
    req = self._make_call(add_params=add_params)
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/query.py", line 290, in _make_call
    raise RequestError(req)
pynautobot.core.query.RequestError: The request failed with code 403 Forbidden: {'detail': 'You do not have permission to perform this action.'}
```

```
14:50:36.015 WARNING django.request :
  Forbidden: /api/extras/tags/
14:50:36.015 WARNING django.server :
  "GET /api/extras/tags/ HTTP/1.1" 403 63
```

```
>>> device_update = device.update({"tags": ["b7ee8585-a088-472a-817e-6e34f836a35f"]})
>>> device_update
True
>>> device.tags
['b7ee8585-a088-472a-817e-6e34f836a35f']
```

```
14:48:10.117 INFO    django.server :
  "PATCH /api/dcim/devices/37be54d8-04c3-4fc8-8179-ee5a8018db2b/ HTTP/1.1" 200 2197
```

Will set up 1.6 LTM instance and re-test but I expect the same behavior there.

**Comment by glennmatthews** [comment]
Same in 1.6:

```
>>> device
<pynautobot.models.dcim.Devices ('rtr1') at 0x10e696280>
>>> device.tags
[]
>>> device.update({"tags": ["cee982b6-94b2-41f2-a546-cbed599859c6"]})
True
>>> nautobot.extras.tags.all()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/endpoint.py", line 111, in all
    return response_loader(req.get(), self.return_obj, self)
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/query.py", line 357, in get
    return req_all()
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/query.py", line 315, in req_all
    req = self._make_call(add_params=add_params)
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/query.py", line 288, in _make_call
    raise RequestError(req)
pynautobot.core.query.RequestError: The request failed with code 403 Forbidden: {'detail': 'You do not have permission to perform this action.'}
```

**Comment by glennmatthews** [comment]
Also an issue in 1.6 for foreign keys, not just M2M:

```
>>> nautobot.dcim.sites.all()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/endpoint.py", line 111, in all
    return response_loader(req.get(), self.return_obj, self)
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/query.py", line 357, in get
    return req_all()
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/query.py", line 315, in req_all
    req = self._make_call(add_params=add_params)
  File "/Users/glennmatthews/Documents/git-stuff/pynautobot/pynautobot/core/query.py", line 288, in _make_call
    raise RequestError(req)
pynautobot.core.query.RequestError: The request failed with code 403 Forbidden: {'detail': 'You do not have permission to perform this action.'}
>>> device.update({"site": {"name": "Site2"}})
True
```

Doesn't *appear* to be an issue in the UI:

![image](https://github.com/user-attachments/assets/4c6e6606-e3c4-40cb-a363-54532e33e532)

![image](https://github.com/user-attachments/assets/87e3a32d-2f0a-4258-b26a-aed60f8b151f)

**Comment by glennmatthews** [comment]
Candidate fix for 2.x:

```diff
diff --git a/nautobot/core/api/mixins.py b/nautobot/core/api/mixins.py
index 8cc9dfff2..26ddb6c57 100644
--- a/nautobot/core/api/mixins.py
+++ b/nautobot/core/api/mixins.py
@@ -109,6 +109,9 @@ class WritableSerializerMixin:
             queryset = self.queryset
         else:
             queryset = self.Meta.model.objects
+        # Apply permissions on accessing the related object(s):
+        if "request" in self.context:
+            queryset = queryset.restrict(self.context["request"].user, "view")
         if isinstance(data, list):
             return [self.get_object(data=entry, queryset=queryset) for entry in data]
         return self.get_object(data=data, queryset=queryset)
```


```
pynautobot.core.query.RequestError: The request failed with code 400 Bad Request: {'location': ["Related object not found using the provided attributes: {'name': 'Site1'}"], 'tags': ["Related object not found using the provided attributes: {'name': 'Grey'}"]}
```

**Comment by lampwins** [comment]
I think we have a good handle the issue now. We were going back and forth on some potential widespread implications to testing but we realized there is already a very similar pattern going on in the UI test cases. While not optimal we can use `EXEMPT_VIEW_PERMISSIONS` in the test cases to avoid that downstream (app ecosystem) test impact, and then better address that concern in a larger permission system enhancement at a later date.

TL;DR: we are good to move forward with this fix now.
